File: java/tree-structures/AVL.java
// Luis Espino 2022

class NodeAVL {
	int value;
	NodeAVL left;
	NodeAVL right;
    int alt;
	
	NodeAVL(int value){
		this.value = value;
		left = null;
		right = null;
        alt = 0;
	}
}


public class AVL {
	NodeAVL root = null;
	
	void add(int value) {
		root = add(value, root);
	}
	
	NodeAVL add(int value, NodeAVL tmp) {
        if (tmp == null) tmp = new NodeAVL(value);
		else if (value < tmp.value) {
            tmp.left = add(value, tmp.left);
            if ((altura(tmp.left)-altura(tmp.right))==2) {
                if (value<tmp.left.value) tmp = srl(tmp);
                else tmp = drl(tmp);
            }            
		}
		else {
            tmp.right = add(value, tmp.right);
            if ((altura(tmp.right)-altura(tmp.left))==2) {
                if (value>tmp.right.value) tmp = srr(tmp);
                else tmp = drr(tmp);
            }            

		}
        int d, i, m;
        d = altura(tmp.right);
        i = altura(tmp.left);
        m = maxi(d,i);
        tmp.alt = m + 1;
        return tmp;
	}

    int altura(NodeAVL tmp) {
        if (tmp == null) return -1;
        else return tmp.alt;
    }

    int maxi(int val1, int val2) {
        return ((val1 > val2) ? val1 : val2);
    }

    NodeAVL srl(NodeAVL t1) {
        NodeAVL t2;
        t2 = t1.left;
        t1.left = t2.right;
        t2.right = t1;
        t1.alt = maxi(altura(t1.left), altura(t1.right))+1;
        t2.alt = maxi(altura(t2.left),t1.alt)+1;
        return t2;
    }

    NodeAVL srr(NodeAVL t1) {
        NodeAVL t2;
        t2 = t1.right;
        t1.right = t2.left;
        t2.left = t1;
        t1.alt = maxi(altura(t1.left), altura(t1.right))+1;
        t2.alt = maxi(altura(t2.right),t1.alt)+1;
        return t2;
    }

    NodeAVL drl(NodeAVL tmp) {
        tmp.left = srr(tmp.left);
        return srl(tmp);
    }

    NodeAVL drr(NodeAVL tmp) {
        tmp.right = srl(tmp.right);
        return srr(tmp);
    }
	
	void preorder(NodeAVL tmp) {
		if (tmp != null) {
			System.out.print(tmp.value+" ");
			preorder(tmp.left);
			preorder(tmp.right);
		}
	}

	void enorder(NodeAVL tmp) {
		if (tmp != null) {			
			enorder(tmp.left);
			System.out.print(tmp.value+" ");
			enorder(tmp.right);
		}
	}
	
	void postorder(NodeAVL tmp) {
		if (tmp != null) {			
			postorder(tmp.left);
			postorder(tmp.right);
			System.out.print(tmp.value+" ");			
		}
	}
	
	public static void main(String[] args) {
		AVL a = new AVL();
        a.add(5);a.add(10);
        a.add(20);a.add(25);
		a.add(30);a.add(35);              
        a.add(40);
		a.preorder(a.root);
		System.out.println();
		a.enorder(a.root);
		System.out.println();
		a.postorder(a.root);
	}   
}

File: cpp/non-linear-structures/avl.cpp
#include <iostream>

class Node
{
public:
    int val;
    Node* izq;
    Node* der;
    int alt;
    Node (int val_) : val(val_), izq(0), der(0), alt(0) {}
};

class AVLTree
{
public:
    Node *raiz;
    AVLTree () : raiz(0) {}
    void add(int val)
    {
        if (raiz!=0) add(val, raiz);
        else raiz = new Node(val);
    }
    void add(int val, Node* &tmp)
    {
        if (tmp==0) tmp = new Node(val);
        else if (val<tmp->val)
        {
            add(val, tmp->izq);
            if ((altura(tmp->izq)-altura(tmp->der))==2)
            {
                if (val<tmp->izq->val) tmp = srl(tmp);
                else tmp = drl(tmp);
            }
        }
        else if (val>tmp->val)
        {
            add(val, tmp->der);
            if ((altura(tmp->der)-altura(tmp->izq))==2)
            {
                if (val>tmp->der->val) tmp = srr(tmp);
                else tmp = drr(tmp);
            }
        }
        int d,i,m;
        d = altura(tmp->der);
        i = altura(tmp->izq);
        m = maxi(d,i);
        tmp->alt = m + 1;
    }
    int altura(Node* tmp)
    {
        if (tmp==0) return -1;
        else return tmp->alt;
    }
    void preorden(Node* tmp)
    {
        if (tmp!=0)
        {
            std::cout << tmp->val << " ";
            preorden(tmp->izq);
            preorden(tmp->der);
        }
    }
    void enorden(Node* tmp)
    {
        if (tmp!=0)
        {
            enorden(tmp->izq);
            std::cout << tmp->val << " ";
            enorden(tmp->der);
        }
    }
    void postorden(Node* tmp)
    {
        if (tmp!=0)
        {
            postorden(tmp->izq);
            postorden(tmp->der);
            std::cout << tmp->val << " ";
        }
    }
    Node* srl(Node* t1)
    {
        Node* t2;
        t2 = t1->izq;
        t1->izq = t2->der;
        t2->der = t1;
        t1->alt = maxi(altura(t1->izq), altura(t1->der))+1;
        t2->alt = maxi(altura(t2->izq),t1->alt)+1;
        return t2;
    }
    Node* srr(Node* t1)
    {
        Node* t2;
        t2 = t1->der;
        t1->der = t2->izq;
        t2->izq = t1;
        t1->alt = maxi(altura(t1->izq), altura(t1->der))+1;
        t2->alt = maxi(altura(t2->der),t1->alt)+1;
        return t2;
    }
    Node* drl(Node* tmp)
    {
        tmp->izq = srr(tmp->izq);
        return srl(tmp);
    }
    Node* drr(Node* tmp)
    {
        tmp->der = srl(tmp->der);
        return srr(tmp);
    }
    int maxi(int val1, int val2)
    {
        return ((val1 > val2) ? val1 : val2);
    }

    int magic(Node* tmp)
    {
        if (tmp!=0)
            return tmp->val+magic(tmp->der)-magic(tmp->izq);
        return 0;
    }
};

int main()
{
    AVLTree* t = new AVLTree();
    //t->add(22);t->add(7);t->add(8);t->add(15);t->add(20);t->add(28);
    //t->add(31);t->add(23);t->add(2);t->add(5);t->add(7);t->add(13);
    //t->add(25);t->add(2);t->add(15);t->add(14);t->add(20);t->add(11);
    //t->add(5);t->add(1);t->add(50);t->add(21);t->add(44);t->add(10);
     //t->add(13);t->add(5);t->add(7);t->add(23);t->add(2);t->add(31);
    //t->add(40);t->add(30);t->add(20);
    //t->add(5);t->add(35);t->add(10);
    //t->add(30);t->add(20);
    //t->add(5); t->add(35); t->add(10);t->add(25);
    //t->add(25);t->add(10);t->add(35); t->add(5);
    //t->add(20); t->add(30);t->add(40);
    t->add(5);t->add(10);t->add(15); t->add(20);
    t->add(25);t->add(30);t->add(35);

    t->preorden(t->raiz);std::cout<<"\n";
    t->enorden(t->raiz);std::cout<<"\n";
    t->postorden(t->raiz);std::cout<<"\n";
    //std::cout << t->altura(t->raiz);
    //std::cout << t->magic(t->raiz);
    return 0;
}